### PR TITLE
ensure hashmap init clears maps 

### LINKED ISSUE #791: freelist hashmap/array inconsistency on page reload
While testing the interface in #786, I've noticed a slight divergence in reload behavior between the array and freelist implementation.

Below testcase showcases the difference very well. TL;DR: we create a freelist with free page ids 5,6,8 and then we create a new one which frees the 5 through 9 in a transaction, marking the whole span as pending.

Now reloading the freelist again from the page that contains 5,6,8 will show different results:

```
func TestFreeList_reload_page_dedupe(t *testing.T) {
	var buf [4096]byte
	f := newTestFreelist()
	f.readIDs([]common.Pgid{5, 6, 8})

	p := (*common.Page)(unsafe.Pointer(&buf[0]))
	if err := f.write(p); err != nil {
		t.Fatal(err)
	}

	f2 := newTestFreelist()
	f2.free(common.Txid(5), common.NewPage(5, common.LeafPageFlag, 0, 4))
	// reload should deduplicate as a pending page when reading from p's freelist
	f2.reload(p)

	if len(f2.getFreePageIDs()) != 0 {
		t.Fatalf("expected empty; got=%v", f2.getFreePageIDs())
	}
	if exp := []common.Pgid{5, 6, 7, 8, 9}; !reflect.DeepEqual(exp, f2.pending[5].ids) {
		t.Fatalf("exp=%v; got=%v", exp, f2.pending[5].ids)
	}
}
```

The array freelist will pass and detect the span as pending, the hashmap will have the overlapping pages pending **and** free.

It stems from the fact that on init, the hashmap will not honor an empty list as an initial state (as in, it will not attempt to initialize anything):
https://github.com/etcd-io/bbolt/blob/main/freelist_hmap.go#L197-L200

Whereas the array will happily take what it's passed and reindex accordingly:
https://github.com/etcd-io/bbolt/blob/main/freelist_array.go#L60-L63

### PR DESCRIPTION
**Pull request by tjungblu**
This reorders some statements in the hashmap initialization to ensure we always start fresh, even when no pageids were passed to it.

fixes #791

--

Might be better to fix this before #775? would make an easier backport.

### CODE REVIEW

File: freelist_test.go [hunk 1]
```diff
@@ -179,6 +179,29 @@ func TestFreelist_releaseRange(t *testing.T) {
 	}
 }
 
+func TestFreeList_reload_page_dedupe(t *testing.T) {
+	var buf [4096]byte
+	f := newTestFreelist()
+	f.readIDs([]common.Pgid{5, 6, 8})
+
+	p := (*common.Page)(unsafe.Pointer(&buf[0]))
```
[the remark this answers is not in the document]
Comment (tjungblu) [reply freelist_test.go:187]: applied

File: freelist_test.go [hunk 2]
```diff
@@ -179,6 +179,29 @@ func TestFreelist_releaseRange(t *testing.T) {
 	}
 }
 
+func TestFreeList_reload_page_dedupe(t *testing.T) {
+	var buf [4096]byte
```
[the remark this answers is not in the document]
Comment (tjungblu) [reply freelist_test.go:183]: applied

File: freelist_test.go [hunk 3]
```diff
@@ -179,6 +179,29 @@ func TestFreelist_releaseRange(t *testing.T) {
...
+	f2.free(common.Txid(5), common.NewPage(5, common.LeafPageFlag, 0, 4))
+	// reload should deduplicate as a pending page when reading from p's freelist
+	f2.reload(p)
+
+	if len(f2.getFreePageIDs()) != 0 {
+		t.Fatalf("expected empty; got=%v", f2.getFreePageIDs())
+	}
+	if exp := []common.Pgid{5, 6, 7, 8, 9}; !reflect.DeepEqual(exp, f2.pending[5].ids) {
+		t.Fatalf("exp=%v; got=%v", exp, f2.pending[5].ids)
+	}
```
[the remark this answers is not in the document]
Comment (tjungblu) [reply freelist_test.go]: Added your first suggestion. As for the second, we should hide the reload behind a `Rollback` as another refactoring step for #789 

As of today, nothing stops you from reloading without the rollback.

[the remark this answers is not in the document]
Comment (tjungblu) [reply freelist_test.go]: Happy to hear other suggestions on how to replicate that inconsistent behavior though, feel free to send an alternative unit test.

[the remark this answers is not in the document]
Comment (ahrtr) [reply freelist_test.go]: > As of today, nothing stops you from reloading without the rollback.

It means the interface isn't well designed. We need to continue to refactor the interface from users perspective. Will think about it.

File: internal/freelist/freelist_test.go [hunk 4]
```diff
@@ -179,6 +179,30 @@ func TestFreelist_releaseRange(t *testing.T) {
...
+	p := common.LoadPage(buf)
+	f.Write(p)
+
+	f2 := newTestFreelist()
+	f.Init([]common.Pgid{})
+
+	f2.Free(common.Txid(5), common.NewPage(5, common.LeafPageFlag, 0, 4))
+
+	// reload should deduplicate as a pending page when reading from p's freelist
+	f2.Reload(p)
```
Comment (ahrtr) [inline internal/freelist/freelist_test.go:196]: When you reload the freelist from a page, it means you already read the freelist from the page in the first place.

**Comment by ahrtr** [comment]
Please see the proposed unit test cases.  We can revisit them when we update the interface later.

```

func TestFreeList_init(t *testing.T) {
	buf := make([]byte, 4096)
	f := newTestFreelist()
	f.Init(common.Pgids{5, 6, 8})

	p := common.LoadPage(buf)
	f.Write(p)

	f2 := newTestFreelist()
	f2.Read(p)
	require.Equal(t, common.Pgids{5, 6, 8}, f2.freePageIds())

	// When initializing the freelist with an empty list of page ID,
	// it should reset the freelist page IDs.
	f2.Init([]common.Pgid{})
	require.Equal(t, common.Pgids{}, f2.freePageIds())
}

func TestFreeList_reload(t *testing.T) {
	buf := make([]byte, 4096)
	f := newTestFreelist()
	f.Init(common.Pgids{5, 6, 8})

	p := common.LoadPage(buf)
	f.Write(p)

	f2 := newTestFreelist()
	f2.Read(p)
	require.Equal(t, common.Pgids{5, 6, 8}, f2.freePageIds())

	f2.Free(common.Txid(5), common.NewPage(10, common.LeafPageFlag, 0, 2))

	// reload shouldn't affect the pending list
	f2.Reload(p)

	require.Equal(t, common.Pgids{5, 6, 8}, f2.freePageIds())
	require.Equal(t, []common.Pgid{10, 11, 12}, f2.pendingPageIds()[5].ids)
}
```

**Comment by tjungblu** [comment]
Works for me, there's still some nil/empty inconsistency:

```
       	Error:      	Not equal: 
        	            	expected: common.Pgids{}
        	            	actual  : common.Pgids(nil)
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,3 +1,2 @@
        	            	-(common.Pgids) {
        	            	-}
        	            	+(common.Pgids) <nil>
```


similar to here, I had to fix it with:
https://github.com/etcd-io/bbolt/pull/786/files#diff-1661bc3663fd3a2f6e7ebc479e8084bd0b1553a456792bdb0d005dbabee969f6R120

**Comment by tjungblu** [comment]
okay that required a bit more empty/nil changes unfortunately. I'll leave those as a separated commit now....

**Pull request merged**